ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -382,6 +386,186 @@ prop_diffusion_dns_can_recover defaultBearerInfo diffScript =
         counterexample (show events)
       $ verify Map.empty 0 (Time 0) (Signal.eventsToList events)
 
+
+-- | A variant of
+-- 'Test.Ouroboros.Network.PeerSelection.prop_governor_target_established_public'
+-- but for running on Diffusion. This means it has to have in consideration the
+-- the logs for all nodes running will all appear in the trace and the test
+-- property should only be valid while a given node is up and running.
+--
+-- We do not need separate above and below variants of this property since it
+-- is not possible to exceed the target.
+--
+prop_diffusion_target_established_public :: AbsBearerInfo
+                                         -> DiffusionScript
+                                         -> Property
+prop_diffusion_target_established_public defaultBearerInfo diffScript =
+    let sim :: forall s . IOSim s Void
+        sim = diffusionSimulation (toBearerInfo defaultBearerInfo)
+                                  diffScript
+                                  tracersExtraWithTimeName
+                                  tracerDiffusionSimWithTimeName
+
+        events :: [Events DiffusionTestTrace]
+        events = fmap ( Signal.eventsFromList
+                      . fmap (\(WithName _ (WithTime t b)) -> (t, b))
+                      )
+               . Trace.toList
+               . splitWithNameTrace
+               . Trace.fromList ()
+               . fmap snd
+               . Trace.toList
+               . fmap (\(WithTime t (WithName name b)) -> (t, WithName name (WithTime t b)))
+               . withTimeNameTraceEvents
+                  @DiffusionTestTrace
+                  @NtNAddr
+               . Trace.fromList (MainReturn (Time 0) () [])
+               . fmap (\(t, tid, tl, te) -> SimEvent t tid tl te)
+               . take 500000
+               . traceEvents
+               $ runSimTrace sim
+
+     in conjoin
+      $ (\ev ->
+        let evsList = eventsToList ev
+            lastTime = fst
+                     . last
+                     $ evsList
+         in classifySimulatedTime lastTime
+          $ classifyNumberOfEvents (length evsList)
+          $ verify_target_established_public ev
+        )
+      <$> events
+  where
+    verify_target_established_public :: Events DiffusionTestTrace -> Property
```

Comment:
Could this be exported by the `Test.Ouroboros.Network.Governor` module.  The same for the other such functions.

Reply:
Not quite because although the code looks pretty much the same we are using different ways of obtaining the trace!

Reply:
I'd extract the core function which does the validation, and export it.  This will allows us to use it in the two different contexts.  But this can be done in a separate PR if you prefer.